File: scripts/LoginValidation.py
# Library to make the verification easier
import re

# Exception for Invalid Login
class InvalidLoginException(Exception):
    pass

# This function was made to verify the data of the login and password. The mensages from the Exceptions are in Brazilian Portuguese language. Because the app is going to be inicially for the brazilian clients 
def user_validation(username, password):
    
    if not username:
        raise InvalidLoginException("O login não pode ser vazio.")
    
    if len(username) > 12:
        raise InvalidLoginException("O login deve ter no máximo 12 caracteres")
    
    for i in range(len(username)):
        
        if username[i] == " ":
            raise InvalidLoginException("O login não pode conter espaços.")
        
        if username[i].isdigit():
            raise InvalidLoginException("O login não pode conter números.")
    
    if len(password) > 128:
        raise InvalidLoginException("A senha deve ter no máximo 128 caracteres.")
    
    if len(password) < 8:
        raise InvalidLoginException("A senha deve ter no mínimo 8 caracteres.")
    
    if not re.search(r'[A-Z]', password):
        raise InvalidLoginException("A senha deve conter pelo menos uma letra maiúscula.")
    
    if not re.search(r'[a-z]', password):
        raise InvalidLoginException("A senha deve conter pelo menos uma letra minúscula.")
    
    if not re.search(r'[0-9]', password):
        raise InvalidLoginException("A senha deve conter pelo menos um número.")
    
    if not re.search(r'[!@#$%^&*(),.?":{}|<>]', password):
        raise InvalidLoginException("A senha deve conter pelo menos um caractere especial.")
    
    if password == username:
        raise InvalidLoginException("A senha não pode ser igual ao nome de usuário.")
    
    return True

File: scripts/LoginValidation.java
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class InvalidLoginException extends Exception {
    public InvalidLoginException(String message) {
        super(message);
    }
}

public class LoginValidation {
    
    public boolean userValidation(String username, String password) throws InvalidLoginException {
        
        // Verificação para null ou strings vazias
        if (username == null || username.isEmpty()) {
            throw new InvalidLoginException("O login não pode ser vazio.");
        }
        
        if (username.length() > 12) {
            throw new InvalidLoginException("O login deve ter no máximo 12 caracteres.");
        }
        
        // Verificar se o login contém espaços ou números
        for (int i = 0; i < username.length(); i++) {
            if (Character.isWhitespace(username.charAt(i))) {
                throw new InvalidLoginException("O login não pode conter espaços.");
            }
            
            if (Character.isDigit(username.charAt(i))) {
                throw new InvalidLoginException("O login não pode conter números.");
            }
        }
        
        if (password == null) {
            throw new InvalidLoginException("A senha não pode ser nula.");
        }
        
        if (password.length() > 128) {
            throw new InvalidLoginException("A senha deve ter no máximo 128 caracteres.");
        }
        
        if (password.length() < 8) {
            throw new InvalidLoginException("A senha deve ter no mínimo 8 caracteres.");
        }
        
        if (!containsUpperCase(password)) {
            throw new InvalidLoginException("A senha deve conter pelo menos uma letra maiúscula.");
        }
        
        if (!containsLowerCase(password)) {
            throw new InvalidLoginException("A senha deve conter pelo menos uma letra minúscula.");
        }
        
        if (!containsDigit(password)) {
            throw new InvalidLoginException("A senha deve conter pelo menos um número.");
        }
        
        if (!containsSpecialCharacter(password)) {
            throw new InvalidLoginException("A senha deve conter pelo menos um caractere especial.");
        }
        
        if (password.equals(username)) {
            throw new InvalidLoginException("A senha não pode ser igual ao nome de usuário.");
        }
        
        return true;
    }
    
    //verificar se a senha contém pelo menos uma letra maiúscula
    private static boolean containsUpperCase(String str) {
        for (char c : str.toCharArray()) {
            if (Character.isUpperCase(c)) {
                return true;
            }
        }
        return false;
    }

    //verificar se a senha contém pelo menos uma letra minúscula
    private static boolean containsLowerCase(String str) {
        for (char c : str.toCharArray()) {
            if (Character.isLowerCase(c)) {
                return true;
            }
        }
        return false;
    }

    //verificar se a senha contém pelo menos um número
    private static boolean containsDigit(String str) {
        for (char c : str.toCharArray()) {
            if (Character.isDigit(c)) {
                return true;
            }
        }
        return false;
    }

    //verificar se a senha contém pelo menos um caractere especial
    private static boolean containsSpecialCharacter(String str) {
        Pattern specialChars = Pattern.compile("[!@#$%^&*(),.?\":{}|<>]");
        Matcher hasSpecial = specialChars.matcher(str);
        return hasSpecial.find();
    }
}
